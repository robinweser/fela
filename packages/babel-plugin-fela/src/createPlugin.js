import generateHash from 'string-hash'
import objectReduce from 'fast-loops/lib/objectReduce'
import arrayReduce from 'fast-loops/lib/arrayReduce'
import arrayEach from 'fast-loops/lib/arrayEach'

const defaultConfig = {
  precompile: true,
  cssProp: true,
}

export default function createPlugin(userConfig = {}) {
  const config = {
    ...defaultConfig,
    ...userConfig,
  }

  return ({ types: t, traverse }) => {
    // helper method to extract static style properties from AST objects
    function extractStaticStyle(props, path) {
      const removeQueue = []

      const staticStyle = objectReduce(
        props,
        (style, node) => {
          const removeCallback = () => props.splice(props.indexOf(node), 1)

          if (
            !node.shorthand &&
            (t.isStringLiteral(node.value) || t.isNumericLiteral(node.value))
          ) {
            removeQueue.push(removeCallback)
            style.push(node)
          } else if (t.isObjectExpression(node.value)) {
            const properties = extractStaticStyle(node.value.properties, path)

            style.push(
              t.objectProperty(node.key, t.objectExpression(properties))
            )

            if (node.value.properties.length === 0) {
              removeQueue.push(removeCallback)
            }
          }

          return style
        },
        []
      )

      arrayEach(removeQueue, cb => cb())
      return staticStyle
    }

    // helper to transform AST ObjectExpressions into JS objects
    function createStaticJSObject(props, path) {
      return arrayReduce(
        props,
        (obj, node) => {
          if (
            !node.shorthand &&
            (t.isStringLiteral(node.value) || t.isNumericLiteral(node.value))
          ) {
            obj[node.key.name] = node.value.value
          } else if (t.isObjectExpression(node.value)) {
            obj[node.key.value] = createStaticJSObject(
              node.value.properties,
              path
            )
          }

          return obj
        },
        {}
      )
    }

    // abstraction to compile arrow functions, function expressions and function declarations
    function getRuleScope(path) {
      const rule = path.node.arguments[0]
      let ruleDeclaration, functionExpression

      if (t.isIdentifier(rule)) {
        if (path.scope.hasBinding(rule.name)) {
          const binded = path.scope.bindings[rule.name].path

          if (
            t.isVariableDeclarator(binded) ||
            t.isFunctionDeclaration(binded)
          ) {
            ruleDeclaration = binded
            functionExpression = t.isVariableDeclarator(binded)
              ? ruleDeclaration.node.init
              : ruleDeclaration.node
          }
        }
      } else if (t.isFunctionDeclaration(rule)) {
        ruleDeclaration = rule
        functionExpression = rule.node
      } else if (
        t.isArrowFunctionExpression(rule) ||
        t.isFunctionExpression(rule)
      ) {
        ruleDeclaration = rule
        functionExpression = rule
      }

      if (t.isArrowFunctionExpression(functionExpression)) {
        if (t.isObjectExpression(functionExpression.body)) {
          functionExpression.body = t.blockStatement([
            t.returnStatement(functionExpression.body),
          ])

          if (t.isVariableDeclarator(ruleDeclaration)) {
            ruleDeclaration.node.init = functionExpression
          } else {
            ruleDeclaration.node = functionExpression
          }
        }
      }

      return {
        ruleDeclaration,
        functionExpression,
      }
    }

    return {
      inherits: require('babel-plugin-syntax-jsx'),
      visitor: {
        CallExpression(path, parentPath) {
          // TODO: check if createComponent is imported from *-fela
          if (path.node.callee.name === 'createComponent') {
            const { ruleDeclaration, functionExpression } = getRuleScope(path)

            if (!ruleDeclaration || !functionExpression) {
              // TODO: Warning?
              return false
            }

            let didTraverse
            let renderer

            // pass renderer as second rule parameter if not already happening
            if (
              functionExpression.params &&
              functionExpression.params.length === 0
            ) {
              functionExpression.params.push(t.identifier('_'))
            }

            if (
              functionExpression.params &&
              functionExpression.params.length === 2
            ) {
              if (!t.isIdentifier(functionExpression.params[1])) {
                // TODO: Warning
                return
              }

              if (functionExpression.params[1].name !== 'renderer') {
                renderer = functionExpression.params[1].name
              }
            }

            if (
              functionExpression.params &&
              functionExpression.params.length === 1
            ) {
              functionExpression.params.push(t.identifier('renderer'))
              renderer = 'renderer'
            }

            const traverser = {
              ObjectExpression(childPath) {
                if (!didTraverse) {
                  childPath.node.root = true
                  didTraverse = true
                }

                if (childPath.node.root) {
                  const props = childPath.node.properties
                  const shouldPrecompile = config.renderer && config.precompile

                  let id, blockBody, className
                  let staticStyle = extractStaticStyle(props, childPath)

                  // static style precompilation
                  if (shouldPrecompile) {
                    const felaRenderer = config.renderer()

                    const jsObject = createStaticJSObject(staticStyle)

                    className = felaRenderer.renderRule(() => jsObject)

                    if (className) {
                      id = className.replace(/ /g, '')

                      const blockBodyItems = [
                        t.expressionStatement(
                          t.assignmentExpression(
                            '=',
                            t.memberExpression(
                              t.memberExpression(
                                t.identifier(renderer),
                                t.identifier('cache')
                              ),
                              t.identifier(id)
                            ),
                            t.objectExpression([
                              t.objectProperty(
                                t.identifier('type'),
                                t.stringLiteral('PRECOMPILATION')
                              ),
                            ])
                          )
                        ),
                      ]

                      // rehydrate all cache elements
                      for (const key in felaRenderer.cache) {
                        const cacheEntry = objectReduce(
                          felaRenderer.cache[key],
                          (entry, value, property) => {
                            entry.push(
                              t.objectProperty(
                                t.identifier(property),
                                t.stringLiteral(value)
                              )
                            )
                            return entry
                          },
                          []
                        )

                        blockBodyItems.push(
                          t.ifStatement(
                            t.unaryExpression(
                              '!',
                              t.memberExpression(
                                t.memberExpression(
                                  t.identifier(renderer),
                                  t.identifier('cache')
                                ),
                                t.stringLiteral(key),
                                true
                              )
                            ),
                            t.expressionStatement(
                              t.assignmentExpression(
                                '=',
                                t.memberExpression(
                                  t.memberExpression(
                                    t.identifier(renderer),
                                    t.identifier('cache')
                                  ),
                                  t.stringLiteral(key),
                                  true
                                ),
                                t.objectExpression(cacheEntry)
                              )
                            )
                          )
                        )
                      }

                      blockBody = t.blockStatement(blockBodyItems)
                    }

                    staticStyle = []
                  }

                  // simple static style prerendering
                  if (staticStyle.length > 0) {
                    id = generateHash(JSON.stringify(staticStyle))

                    blockBody = t.blockStatement([
                      t.expressionStatement(
                        t.assignmentExpression(
                          '=',
                          t.memberExpression(
                            t.memberExpression(
                              t.identifier(renderer),
                              t.identifier('cache')
                            ),
                            t.identifier(id)
                          ),
                          t.objectExpression([
                            t.objectProperty(
                              t.identifier('type'),
                              t.stringLiteral('PRERENDERING')
                            ),
                            t.objectProperty(
                              t.identifier('className'),
                              t.callExpression(
                                t.memberExpression(
                                  t.identifier(renderer),
                                  t.identifier('renderRule')
                                ),
                                [
                                  t.ArrowFunctionExpression(
                                    [],
                                    t.objectExpression(staticStyle)
                                  ),
                                ]
                              )
                            ),
                          ])
                        )
                      ),
                    ])
                  }

                  if (id && blockBody) {
                    if (shouldPrecompile) {
                      childPath.node.properties.unshift(
                        t.objectProperty(
                          t.identifier('_className'),
                          t.stringLiteral(className)
                        )
                      )
                    } else {
                      childPath.node.properties.unshift(
                        t.objectProperty(
                          t.identifier('_className'),
                          t.memberExpression(
                            t.memberExpression(
                              t.memberExpression(
                                t.identifier(renderer),
                                t.identifier('cache')
                              ),
                              t.identifier(id)
                            ),
                            t.identifier('className')
                          )
                        )
                      )
                    }

                    let isRoot

                    traverse(
                      functionExpression,
                      {
                        BlockStatement(bPath) {
                          if (!isRoot) {
                            isRoot = true

                            bPath.node.body.unshift(
                              t.ifStatement(
                                t.unaryExpression(
                                  '!',
                                  t.memberExpression(
                                    t.memberExpression(
                                      t.identifier(renderer),
                                      t.identifier('cache')
                                    ),
                                    t.identifier(id)
                                  )
                                ),
                                blockBody
                              )
                            )
                          }
                        },
                      },
                      childPath.scope,
                      childPath
                    )
                  }
                }
              },
            }

            if (ruleDeclaration.traverse) {
              ruleDeclaration.traverse(traverser)
            } else {
              traverse(ruleDeclaration, traverser, path.scope, path)
            }
          }
        },
        JSXAttribute(path) {
          if (!config.cssProp) {
            return false
          }

          if (
            path.node.name.name === 'css' &&
            t.isJSXExpressionContainer(path.node.value)
          ) {
            path.node.name.name = 'className'
            path.node.value.expression = t.callExpression(
              t.memberExpression(
                t.identifier('renderer'),
                t.identifier('renderRule')
              ),
              [t.arrowFunctionExpression([], path.node.value.expression)]
            )
          }
        },
      },
    }
  }
}
