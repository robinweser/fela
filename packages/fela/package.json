{
  "name": "fela",
  "version": "11.4.0",
  "description": "State-Driven Styling in JavaScript",
  "typings": "index.d.ts",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "sideEffects": false,
  "scripts": {
    "build:umd": "webpack --config webpack.config.js"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.d.ts",
    "lib/**",
    "es/**",
    "umd/**"
  ],
  "repository": "https://github.com/robinweser/fela/",
  "keywords": [
    "fela",
    "dynamic styling",
    "stylesheet",
    "css",
    "styling",
    "functional styling",
    "functional css",
    "atomic css",
    "state-driven",
    "atomic design",
    "state-driven styling",
    "cssinjs"
  ],
  "author": "Robin Weser",
  "license": "MIT",
  "dependencies": {
    "css-in-js-utils": "^3.0.0",
    "csstype": "^2.5.5",
    "fast-loops": "^1.0.0",
    "fela-utils": "^11.4.0",
    "isobject": "^3.0.1"
  },
  "gitHead": "4cc4686d9e16877251ac5c73e161076ec7c113a1"
}
