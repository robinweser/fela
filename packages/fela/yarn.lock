# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


css-in-js-utils@2.0.0, css-in-js-utils@^2.0.0:
  version "2.0.0"
  resolved "https://registry.yarnpkg.com/css-in-js-utils/-/css-in-js-utils-2.0.0.tgz#5af1dd70f4b06b331f48d22a3d86e0786c0b9435"
  dependencies:
    hyphenate-style-name "^1.0.2"

fast-loops@^1.0.0:
  version "1.0.0"
  resolved "https://registry.yarnpkg.com/fast-loops/-/fast-loops-1.0.0.tgz#2cdd7e0ff67343b2b5f5e627d855a50b4bed559a"

fela-tools@^5.1.2:
  version "5.1.2"
  resolved "https://registry.yarnpkg.com/fela-tools/-/fela-tools-5.1.2.tgz#c6760b42d72116911c2c0341b6d880cfc11c76a8"
  dependencies:
    css-in-js-utils "2.0.0"
    fast-loops "^1.0.0"
    fela "^6.1.3"
    fela-utils "^8.0.3"

fela-utils@^8.0.3:
  version "8.0.3"
  resolved "https://registry.yarnpkg.com/fela-utils/-/fela-utils-8.0.3.tgz#4a1728e7f5711e26f6f37da9c85a6b270c84c466"
  dependencies:
    css-in-js-utils "^2.0.0"
    fast-loops "^1.0.0"

fela@^6.1.3:
  version "6.1.3"
  resolved "https://registry.yarnpkg.com/fela/-/fela-6.1.3.tgz#c3bc8726dada63c23914bed9ca79d902e61e9a74"
  dependencies:
    css-in-js-utils "2.0.0"
    fast-loops "^1.0.0"
    fela-utils "^8.0.3"

hyphenate-style-name@^1.0.2:
  version "1.0.2"
  resolved "https://registry.yarnpkg.com/hyphenate-style-name/-/hyphenate-style-name-1.0.2.tgz#31160a36930adaf1fc04c6074f7eb41465d4ec4b"

string-hash@^1.1.3:
  version "1.1.3"
  resolved "https://registry.yarnpkg.com/string-hash/-/string-hash-1.1.3.tgz#e8aafc0ac1855b4666929ed7dd1275df5d6c811b"
