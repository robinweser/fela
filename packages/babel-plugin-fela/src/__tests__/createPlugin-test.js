import path from 'path'
import fs from 'fs'

import { transformFileSync } from 'babel-core'
import { createRenderer } from 'fela'
import webPreset from 'fela-preset-web'

import createPlugin from '../createPlugin'

const fixturePath = '/__fixtures__/'

function getFixtures(folder) {
  return fs
    .readdirSync(path.join(__dirname, fixturePath, folder))
    .reduce((fixtures, file) => {
      fixtures.push(path.join(folder, file))
      return fixtures
    }, [])
}

function transformFile(filename, plugin) {
  const filePath = `${fixturePath}${filename}`

  return [
    filePath,
    transformFileSync(path.join(__dirname, filePath), {
      plugins: [plugin],
      babelrc: false,
    }).code,
  ]
}

describe('Using babel-plugin-fela', () => {
  it('should prerender static styles as a separate rule', () => {
    const plugin = createPlugin()

    getFixtures('createComponent').forEach(fixture => {
      expect(transformFile(fixture, plugin)).toMatchSnapshot()
    })
  })

  it('should precompile static styles', () => {
    const plugin = createPlugin({
      precompile: true,
      renderer: () =>
        createRenderer({
          plugins: [...webPreset],
        }),
    })

    getFixtures('createComponent').forEach(fixture => {
      expect(transformFile(fixture, plugin)).toMatchSnapshot()
    })
  })

  it('should transform inline css props on HTML nodes', () => {
    const plugin = createPlugin({
      cssProp: true
    })

    getFixtures('css-prop').forEach(fixture => {
      expect(transformFile(fixture, plugin)).toMatchSnapshot()
    })
  })
})
