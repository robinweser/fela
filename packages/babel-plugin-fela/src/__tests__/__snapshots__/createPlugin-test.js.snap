// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Using babel-plugin-fela should precompile static styles 1`] = `
Array [
  "/__fixtures__/createComponent/createComponent.js",
  "import { createComponent } from 'react-fela';

const rule = ({ fontSize, padding }, renderer) => {
  if (!renderer.cache.abcd) {
    renderer.cache.abcd = {
      type: 'PRECOMPILATION'
    };
    if (!renderer.cache['colorred']) renderer.cache['colorred'] = {
      type: 'RULE',
      className: 'a',
      selector: '.a',
      declaration: 'color:red',
      media: '',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)backgroundColorred']) renderer.cache['(min-height: 300px)backgroundColorred'] = {
      type: 'RULE',
      className: 'b',
      selector: '.b',
      declaration: 'background-color:red',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)lineHeight2']) renderer.cache['(min-height: 300px)lineHeight2'] = {
      type: 'RULE',
      className: 'c',
      selector: '.c',
      declaration: 'line-height:2',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px):hovercolorblack']) renderer.cache['(min-height: 300px):hovercolorblack'] = {
      type: 'RULE',
      className: 'd',
      selector: '.d:hover',
      declaration: 'color:black',
      media: '(min-height: 300px)',
      support: ''
    };
  }

  return {
    _className: 'a b c d',

    fontSize,

    '@media (min-height: 300px)': {
      ':hover': {
        paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should precompile static styles 2`] = `
Array [
  "/__fixtures__/createComponent/createComponentEmbedded.js",
  "import { createComponent } from 'react-fela';

export default createComponent(({ fontSize, padding }, renderer) => {
  if (!renderer.cache.abcd) {
    renderer.cache.abcd = {
      type: 'PRECOMPILATION'
    };
    if (!renderer.cache['colorred']) renderer.cache['colorred'] = {
      type: 'RULE',
      className: 'a',
      selector: '.a',
      declaration: 'color:red',
      media: '',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)backgroundColorred']) renderer.cache['(min-height: 300px)backgroundColorred'] = {
      type: 'RULE',
      className: 'b',
      selector: '.b',
      declaration: 'background-color:red',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)lineHeight2']) renderer.cache['(min-height: 300px)lineHeight2'] = {
      type: 'RULE',
      className: 'c',
      selector: '.c',
      declaration: 'line-height:2',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px):hovercolorblack']) renderer.cache['(min-height: 300px):hovercolorblack'] = {
      type: 'RULE',
      className: 'd',
      selector: '.d:hover',
      declaration: 'color:black',
      media: '(min-height: 300px)',
      support: ''
    };
  }

  return {
    _className: 'a b c d',

    fontSize,

    '@media (min-height: 300px)': {
      ':hover': {
        paddingLeft: padding
      }
    }
  };
});",
]
`;

exports[`Using babel-plugin-fela should precompile static styles 3`] = `
Array [
  "/__fixtures__/createComponent/createComponentEmptyParameter.js",
  "import { createComponent } from 'react-fela';

const rule = (_, renderer) => {
  if (!renderer.cache.abcd) {
    renderer.cache.abcd = {
      type: 'PRECOMPILATION'
    };
    if (!renderer.cache['colorred']) renderer.cache['colorred'] = {
      type: 'RULE',
      className: 'a',
      selector: '.a',
      declaration: 'color:red',
      media: '',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)backgroundColorred']) renderer.cache['(min-height: 300px)backgroundColorred'] = {
      type: 'RULE',
      className: 'b',
      selector: '.b',
      declaration: 'background-color:red',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)lineHeight2']) renderer.cache['(min-height: 300px)lineHeight2'] = {
      type: 'RULE',
      className: 'c',
      selector: '.c',
      declaration: 'line-height:2',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px):hovercolorblack']) renderer.cache['(min-height: 300px):hovercolorblack'] = {
      type: 'RULE',
      className: 'd',
      selector: '.d:hover',
      declaration: 'color:black',
      media: '(min-height: 300px)',
      support: ''
    };
  }

  return {
    _className: 'a b c d'
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should precompile static styles 4`] = `
Array [
  "/__fixtures__/createComponent/createComponentFunctionExpression.js",
  "import { createComponent } from 'react-fela';

function rule({ fontSize, padding }, renderer) {
  if (!renderer.cache.abcd) {
    renderer.cache.abcd = {
      type: 'PRECOMPILATION'
    };
    if (!renderer.cache['colorred']) renderer.cache['colorred'] = {
      type: 'RULE',
      className: 'a',
      selector: '.a',
      declaration: 'color:red',
      media: '',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)backgroundColorred']) renderer.cache['(min-height: 300px)backgroundColorred'] = {
      type: 'RULE',
      className: 'b',
      selector: '.b',
      declaration: 'background-color:red',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)lineHeight2']) renderer.cache['(min-height: 300px)lineHeight2'] = {
      type: 'RULE',
      className: 'c',
      selector: '.c',
      declaration: 'line-height:2',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px):hovercolorblack']) renderer.cache['(min-height: 300px):hovercolorblack'] = {
      type: 'RULE',
      className: 'd',
      selector: '.d:hover',
      declaration: 'color:black',
      media: '(min-height: 300px)',
      support: ''
    };
  }

  return {
    _className: 'a b c d',

    fontSize,

    '@media (min-height: 300px)': {
      ':hover': {
        paddingLeft: padding
      }
    }
  };
}

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should precompile static styles 5`] = `
Array [
  "/__fixtures__/createComponent/createComponentFunctionVariable.js",
  "import { createComponent } from 'react-fela';

const rule = function ({ fontSize, padding }, renderer) {
  if (!renderer.cache.abcd) {
    renderer.cache.abcd = {
      type: 'PRECOMPILATION'
    };
    if (!renderer.cache['colorred']) renderer.cache['colorred'] = {
      type: 'RULE',
      className: 'a',
      selector: '.a',
      declaration: 'color:red',
      media: '',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)backgroundColorred']) renderer.cache['(min-height: 300px)backgroundColorred'] = {
      type: 'RULE',
      className: 'b',
      selector: '.b',
      declaration: 'background-color:red',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)lineHeight2']) renderer.cache['(min-height: 300px)lineHeight2'] = {
      type: 'RULE',
      className: 'c',
      selector: '.c',
      declaration: 'line-height:2',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px):hovercolorblack']) renderer.cache['(min-height: 300px):hovercolorblack'] = {
      type: 'RULE',
      className: 'd',
      selector: '.d:hover',
      declaration: 'color:black',
      media: '(min-height: 300px)',
      support: ''
    };
  }

  return {
    _className: 'a b c d',

    fontSize,

    '@media (min-height: 300px)': {
      ':hover': {
        paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should precompile static styles 6`] = `
Array [
  "/__fixtures__/createComponent/createComponentInvalidRenderer.js",
  "import { createComponent } from 'react-fela';

const rule = ({ fontSize, padding }, { foo }) => {
  return {
    fontSize,
    color: 'red',
    '@media (min-height: 300px)': {
      backgroundColor: 'red',
      lineHeight: 2,
      ':hover': {
        color: 'black',
        paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should precompile static styles 7`] = `
Array [
  "/__fixtures__/createComponent/createComponentMergeClassName.js",
  "import { createComponent } from 'react-fela';

const rule = ({ fontSize, padding }, renderer) => {
  if (!renderer.cache.foobarabcd) {
    renderer.cache.foobarabcd = {
      type: 'PRECOMPILATION'
    };
    if (!renderer.cache['colorred']) renderer.cache['colorred'] = {
      type: 'RULE',
      className: 'a',
      selector: '.a',
      declaration: 'color:red',
      media: '',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)backgroundColorred']) renderer.cache['(min-height: 300px)backgroundColorred'] = {
      type: 'RULE',
      className: 'b',
      selector: '.b',
      declaration: 'background-color:red',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px)lineHeight2']) renderer.cache['(min-height: 300px)lineHeight2'] = {
      type: 'RULE',
      className: 'c',
      selector: '.c',
      declaration: 'line-height:2',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!renderer.cache['(min-height: 300px):hovercolorblack']) renderer.cache['(min-height: 300px):hovercolorblack'] = {
      type: 'RULE',
      className: 'd',
      selector: '.d:hover',
      declaration: 'color:black',
      media: '(min-height: 300px)',
      support: ''
    };
  }

  return {
    _className: 'foo bar a b c d',

    fontSize,

    '@media (min-height: 300px)': {
      ':hover': {
        paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should precompile static styles 8`] = `
Array [
  "/__fixtures__/createComponent/createComponentRendererReference.js",
  "import { createComponent } from 'react-fela';

const rule = ({ fontSize, padding }, felaRenderer) => {
  if (!felaRenderer.cache.abcd) {
    felaRenderer.cache.abcd = {
      type: 'PRECOMPILATION'
    };
    if (!felaRenderer.cache['colorred']) felaRenderer.cache['colorred'] = {
      type: 'RULE',
      className: 'a',
      selector: '.a',
      declaration: 'color:red',
      media: '',
      support: ''
    };
    if (!felaRenderer.cache['(min-height: 300px)backgroundColorred']) felaRenderer.cache['(min-height: 300px)backgroundColorred'] = {
      type: 'RULE',
      className: 'b',
      selector: '.b',
      declaration: 'background-color:red',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!felaRenderer.cache['(min-height: 300px)lineHeight2']) felaRenderer.cache['(min-height: 300px)lineHeight2'] = {
      type: 'RULE',
      className: 'c',
      selector: '.c',
      declaration: 'line-height:2',
      media: '(min-height: 300px)',
      support: ''
    };
    if (!felaRenderer.cache['(min-height: 300px):hovercolorblack']) felaRenderer.cache['(min-height: 300px):hovercolorblack'] = {
      type: 'RULE',
      className: 'd',
      selector: '.d:hover',
      declaration: 'color:black',
      media: '(min-height: 300px)',
      support: ''
    };
  }

  return {
    _className: 'a b c d',

    fontSize,

    '@media (min-height: 300px)': {
      ':hover': {
        paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should prerender static styles as a separate rule 1`] = `
Array [
  "/__fixtures__/createComponent/createComponent.js",
  "import { createComponent } from 'react-fela';

const rule = ({ fontSize, padding }, renderer) => {
  if (!renderer.cache.1258790278) {
    renderer.cache.1258790278= {
      type: 'PRERENDERING',
      className: renderer.renderRule(() => ({
        color: 'red',
        '@media (min-height: 300px)': {
          backgroundColor: 'red',
          lineHeight: 2,
          ':hover': {
            color: 'black'
          }
        }
      }))
    };
  }

  return {
    _className: renderer.cache.1258790278.className,

    fontSize, '@media (min-height: 300px)': { ':hover': { paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should prerender static styles as a separate rule 2`] = `
Array [
  "/__fixtures__/createComponent/createComponentEmbedded.js",
  "import { createComponent } from 'react-fela';

export default createComponent(({ fontSize, padding }, renderer) => {
  if (!renderer.cache.3434345636) {
    renderer.cache.3434345636= {
      type: 'PRERENDERING',
      className: renderer.renderRule(() => ({
        color: 'red',
        '@media (min-height: 300px)': {
          backgroundColor: 'red',
          lineHeight: 2,
          ':hover': {
            color: 'black'
          }
        }
      }))
    };
  }

  return {
    _className: renderer.cache.3434345636.className,

    fontSize, '@media (min-height: 300px)': { ':hover': { paddingLeft: padding
      }
    }
  };
});",
]
`;

exports[`Using babel-plugin-fela should prerender static styles as a separate rule 3`] = `
Array [
  "/__fixtures__/createComponent/createComponentEmptyParameter.js",
  "import { createComponent } from 'react-fela';

const rule = (_, renderer) => {
  if (!renderer.cache.3904749910) {
    renderer.cache.3904749910= {
      type: 'PRERENDERING',
      className: renderer.renderRule(() => ({
        color: 'red',
        '@media (min-height: 300px)': {
          backgroundColor: 'red',
          lineHeight: 2,
          ':hover': {
            color: 'black'
          }
        }
      }))
    };
  }

  return {
    _className: renderer.cache.3904749910.className
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should prerender static styles as a separate rule 4`] = `
Array [
  "/__fixtures__/createComponent/createComponentFunctionExpression.js",
  "import { createComponent } from 'react-fela';

function rule({ fontSize, padding }, renderer) {
  if (!renderer.cache.1994830291) {
    renderer.cache.1994830291= {
      type: 'PRERENDERING',
      className: renderer.renderRule(() => ({
        color: 'red',
        '@media (min-height: 300px)': {
          backgroundColor: 'red',
          lineHeight: 2,
          ':hover': {
            color: 'black'
          }
        }
      }))
    };
  }

  return {
    _className: renderer.cache.1994830291.className,

    fontSize, '@media (min-height: 300px)': { ':hover': { paddingLeft: padding
      }
    }
  };
}

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should prerender static styles as a separate rule 5`] = `
Array [
  "/__fixtures__/createComponent/createComponentFunctionVariable.js",
  "import { createComponent } from 'react-fela';

const rule = function ({ fontSize, padding }, renderer) {
  if (!renderer.cache.1173747665) {
    renderer.cache.1173747665= {
      type: 'PRERENDERING',
      className: renderer.renderRule(() => ({
        color: 'red',
        '@media (min-height: 300px)': {
          backgroundColor: 'red',
          lineHeight: 2,
          ':hover': {
            color: 'black'
          }
        }
      }))
    };
  }

  return {
    _className: renderer.cache.1173747665.className,

    fontSize, '@media (min-height: 300px)': { ':hover': { paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should prerender static styles as a separate rule 6`] = `
Array [
  "/__fixtures__/createComponent/createComponentInvalidRenderer.js",
  "import { createComponent } from 'react-fela';

const rule = ({ fontSize, padding }, { foo }) => {
  return {
    fontSize,
    color: 'red',
    '@media (min-height: 300px)': {
      backgroundColor: 'red',
      lineHeight: 2,
      ':hover': {
        color: 'black',
        paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should prerender static styles as a separate rule 7`] = `
Array [
  "/__fixtures__/createComponent/createComponentMergeClassName.js",
  "import { createComponent } from 'react-fela';

const rule = ({ fontSize, padding }, renderer) => {
  if (!renderer.cache.2598482558) {
    renderer.cache.2598482558= {
      type: 'PRERENDERING',
      className: renderer.renderRule(() => ({
        _className: 'foo bar',

        color: 'red',
        '@media (min-height: 300px)': {
          backgroundColor: 'red',
          lineHeight: 2,
          ':hover': {
            color: 'black'
          }
        }
      }))
    };
  }

  return {
    _className: renderer.cache.2598482558.className,
    fontSize, '@media (min-height: 300px)': { ':hover': { paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should prerender static styles as a separate rule 8`] = `
Array [
  "/__fixtures__/createComponent/createComponentRendererReference.js",
  "import { createComponent } from 'react-fela';

const rule = ({ fontSize, padding }, felaRenderer) => {
  if (!felaRenderer.cache.3309150636) {
    felaRenderer.cache.3309150636= {
      type: 'PRERENDERING',
      className: felaRenderer.renderRule(() => ({
        color: 'red',
        '@media (min-height: 300px)': {
          backgroundColor: 'red',
          lineHeight: 2,
          ':hover': {
            color: 'black'
          }
        }
      }))
    };
  }

  return {
    _className: felaRenderer.cache.3309150636.className,

    fontSize, '@media (min-height: 300px)': { ':hover': { paddingLeft: padding
      }
    }
  };
};

export default createComponent(rule);",
]
`;

exports[`Using babel-plugin-fela should transform inline css props on HTML nodes 1`] = `
Array [
  "/__fixtures__/css-prop/ArrowFunctionComponent.js",
  "const Button = () => <div className={renderer.renderRule(() => ({
  color: 'red',
  fontSize: 12
}))}>
    Hello
  </div>;",
]
`;

exports[`Using babel-plugin-fela should transform inline css props on HTML nodes 2`] = `
Array [
  "/__fixtures__/css-prop/ClassComponent.js",
  "export default class Button {
  render() {
    return <div className={renderer.renderRule(() => ({
      color: 'red',
      fontSize: 12
    }))}>
        Hello
      </div>;
  }
}",
]
`;

exports[`Using babel-plugin-fela should transform inline css props on HTML nodes 3`] = `
Array [
  "/__fixtures__/css-prop/FunctionComponent.js",
  "export default function Button() {
  return <div className={renderer.renderRule(() => ({
    color: 'red',
    fontSize: 12
  }))}>
      Hello
    </div>;
}",
]
`;
